[Heartbeat] Dedupe screenshots / Extra Args

andrewvc opened the pull request:
Counterpart of https://github.com/elastic/synthetics/pull/282
Fixes: https://github.com/elastic/beats/issues/26188

Adds support for screenshot blocks as mentioned in the synthetics PR, and also switches over to using the new flags / capabilities system. This is done by using `--rich-events` by default as a synthetics flag, and also exposing a new `synthetics_args` option for synthetics, allowing users to use configs like:

```yaml
- type: browser
  enabled: true
  id: my-monitor
  name: My Monitor
  synthetics_args: ["--capability", "trace", "metrics", "filmstrips", "ssblocks"]
  source:
    inline:
      script:
        step("load homepage", async () => {
            await page.goto('https://www.elastic.co');
        });
        step("hover over products menu", async () => {
            await page.hover('css=[data-nav-item=products]');
        });
        step("failme", async () => {
            await page.hhover('css=[data-nav-item=products]');
        });
  schedule: "@every 1m"
```

This PR also moves some of the logic from `browser.go` to `suite.go`, which was also renamed to be cleaner and more in keeping with idiomatic go. This also made adding a small test possible. We still need a better story around E2E testing, but for now this will suffice.

Finally, this adds mappings for `screenshot_ref` objects as well as blocks, and routes them to the correct dataset. It also improves tests for enrichment of events to help test this.

elasticmachine commented:
Pinging @elastic/uptime (Team:Uptime)

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-25808/detail/PR-25808/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-25808/detail/PR-25808/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-25808/detail/PR-25808/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-25808/detail/PR-25808/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_25808.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #25808 opened`



* Start Time: 2021-05-20T22:11:23.070+0000


* Duration: 59 min 43 sec


* Commit: 91fcabfbdcfbd7e1089f1ba5ce09ceb2f35660dd



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 69  |
| Skipped      | 0 |
| Total        | 69   |


#### Trends :test_tube:


![Image of Build Times](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-25808/buildTimeGraph/png)


![Image of Tests](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-25808/1/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :green_heart: Flaky test report
Tests succeeded.


<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

### Test stats :test_tube:
| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 69  |
| Skipped      | 0 |
| Total        | 69   |



</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

vigneshshanmugam reviewed (commented):
LGTM, tested and it works as expected. 

We might need to change one more thing to get it merged. adding `step/block` to the screenshot datastream index.

andrewvc commented:
PR now adds mapping and properly routes to correct dataset. Still needs tests

vigneshshanmugam commented on heartbeat/_meta/fields.common.yml:
```diff
@@ -174,6 +174,22 @@
+            - name: blocks
+              type: group
+              fields: 
+                - name: hash
```
Need to add `top` and `left` here.

andrewvc replied:
Good catch!

vigneshshanmugam replied:
I don't see the `top` & `left` yet in the new changes.

vigneshshanmugam commented on heartbeat/_meta/fields.common.yml:
```diff
@@ -174,6 +174,22 @@
+          fields:
+            - name: width
+              type: integer
+              description: Width of the full screenshot in pixels.
```
Shall we remove the use of full here? Full seems like full page screenshot but we are capture only the viewport width and height.

andrewvc replied:
Hmmm, I intended it to mean not of an individual block. I'll clarify that.

vigneshshanmugam commented on x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:
```diff
@@ -108,23 +109,31 @@ func runCmd(
 	}
 
+	// Default capabilities
+	capabilities = append(capabilities, "trace", "metrics", "filmstrips", "ssblocks")
```
I am not sure if its the right way to do it. 

I was thinking we should make it configurable by introducing Capability flag in the `config.go` in addition to params and then using it here for experimentation. 

moving forward once the feature is built in to the UI, we can add it to `--rich-events` in synthetics. 

WDYT?

andrewvc replied:
Ah, I see what you mean. That makes sense, will make those changes!

vigneshshanmugam commented on x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:
```diff
@@ -168,9 +177,6 @@ func runCmd(
 	// Close mpx after the process is done and all events have been sent / consumed
 	go func() {
 		err := cmd.Wait()
-		if err != nil {
```
Do we do anything special now for handling errors messages?

vigneshshanmugam replied:
Sure, thanks for the info.

andrewvc commented on x-pack/heartbeat/monitors/browser/suite.go:
```diff
@@ -0,0 +1,108 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package browser
```
This file isn't new, it's just moved. Enough of it changed due to renaming the `ss` var to `s` that git thinks it's all new.

vigneshshanmugam commented on CHANGELOG.next.asciidoc:
```diff
@@ -106,6 +106,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Change logging in logs input to structure logging. Some log message formats have changed. {pull}25299[25299]
 
 *Heartbeat*
+- Add support for screenshot blocks and use newer synthetics flags that only worker in newer synthetics betas. {pull}25808[25808]
```
```suggestion
- Add support for screenshot blocks and use newer synthetics flags that only works in newer synthetics betas. {pull}25808[25808]
```

vigneshshanmugam commented on x-pack/heartbeat/monitors/browser/suite_test.go:
```diff
@@ -114,3 +115,37 @@ func TestEmptySource(t *testing.T) {
+		},
+		{
+			"kitchen sink",
+			&Config{SyntheticsArgs: []string{"--capability", "trace"}, Sandbox: true},
```
I am curios to see a test with capability > 2 and other flags as part of synthetic args. Reason being variadic args are prone to errors.

vigneshshanmugam reviewed (approved):
LGTM after the `top` and `left` changes.

andrewvc commented:
@vigneshshanmugam I believe I've addressed your concerns. I left out `top` and `left` on the mapping intentionally. We'll probably never query them, so it makes sense to save a field. I've added some notes to the fields yml indicating why this is.

vigneshshanmugam commented:
@andrewvc Spot on, Sounds good to me ++.

andrewvc commented:
@Mergifyio backport 7.x

andrewvc commented:
@Mergifyio backport 7.14
